Implemented VK sync plugins

antseppa opened the pull request.

Commit:
[sociald] Initial VK support including VK Posts sync adaptor. Contributes to JB#19270

This commit adds VK as a supported social network in sociald, and
adds a posts sync adaptor.

Contributes to JB#19270

Commit:
[sociald] Add VK contacts sync adaptor. Contributes to JB#19272

This commit adds (one-way) contacts sync adaptor for VKontakte.

Contributes to JB#19272

Commit:
[sociald] Add VK calendar event integration. Contributes to JB#19273

This commit adds a sync adaptor which loads event data from VK and
caches it in the local device database.

Contributes to JB#19273

Commit:
[sociald] Add VK notifications sync adaptor. Contributes to JB#19271

This commit adds a VK notifications sync adaptor.

Contributes to JB#19271

Commit:
[sociald] Add VK images sync adaptor. Contributes to JB#19274

This commit adds a sync adapter to retrieve photo metadata from VK
for the user.  This data is added to the libsocialcache vk images
database, to be displayed in the gallery application.

Contributes to JB#19274

Commit:
Make VK post identifier unique

Commit:
Increased VK image resolution

Tried different values but anything less than this will seem blurred
even on Jolla 1.

Commit:
Select image resolution based on screen dimensions

Commit:
Only sync new VK posts

Commit:
Manage social image cache

Commit:
Handle copyPost fields

Commit:
Save all copyPost images

Commit:
[buteo-sync-plugins-social] Handle VK photo album posts. Contributes to MER#1288

chriadam commented on src/vk/vk-contacts/vkcontactsyncadaptor.cpp:
```diff
@@ -0,0 +1,1010 @@
+
+void VKContactSyncAdaptor::sync(const QString &dataTypeString, int accountId)
+{
+    m_apiRequestsRemaining[accountId] = 99; // assume we can make up to 99 requests per sync, before being throttled.
```
I think this is incorrect.  I think the limits are MUCH smaller than this, although I cannot remember what the limits are at the moment.

antseppa replied:
The API documentation says two per second and 5000 per hour https://vk.com/pages?oid=-17680044&p=Limitations_for_Calling_Ads_API_Methods

chriadam commented on src/vk/vk-contacts/vkcontactsyncadaptor.cpp:
```diff
@@ -0,0 +1,1010 @@
+    }
+
+    // parse the remote contact information from the response
+    QString continuationUrl;
```
this continuationUrl is never set, we never parse it.  I guess when I originally wrote this, I wasn't able to test with an account that had more than about 4 friends... We need to figure out what this should be, and parse it properly.

antseppa replied:
I have the same problem, need to test this with someone who has enough friends in VK.

chriadam commented on src/vk/vk-contacts/vkcontactsyncadaptor.cpp:
```diff
@@ -0,0 +1,1010 @@
+
+    // parse the remote contact information from the response
+    QString continuationUrl;
+    QJsonObject obj = QJsonDocument::fromJson(data).object(); // XXX TODO: check errors / continuations / etc.
```
Also, error checking here would be good too ;-)

chriadam commented on src/vk/vk-notifications/vknotificationsyncadaptor.cpp:
```diff
@@ -0,0 +1,150 @@
+        SOCIALD_LOG_DEBUG("sync aborted, skipping finalize of VK Notifications from account:" << accountId);
+    } else {
+        SOCIALD_LOG_DEBUG("finalizing VK Notifications sync with account:" << accountId);
+        //m_db.purgeOldNotifications(OLD_NOTIFICATION_LIMIT_IN_DAYS); // TODO
```
is this still TODO or has it been implemented some other way?

chriadam commented on src/vk/vk-posts/vk-posts.pri:
```diff
@@ -0,0 +1,5 @@
+CONFIG += link_pkgconfig meegotouchevents-qt5
```
is meegotouchevents-qt5 still a thing?

antseppa replied:
a leftover, I'll remove it.

chriadam commented on src/vk/vk-posts/vkpostsyncadaptor.cpp:
```diff
@@ -0,0 +1,472 @@
+                    SOCIALD_LOG_DEBUG("TODO: unhandled newsfeed item type:" << object.value(QStringLiteral("type")).toString() << ", skipping.");
+                } else if (object.value(QStringLiteral("type")).toString() == QStringLiteral("wall_photo")) {
+                    SOCIALD_LOG_DEBUG("TODO: unhandled newsfeed item type:" << object.value(QStringLiteral("type")).toString() << ", skipping.");
+                } else if (object.value(QStringLiteral("type")).toString() == QStringLiteral("note")) {
```
we should probably handle note and wall_photo post types if possible.  It shouldn't be too tricky, I hope?

antseppa replied:
true, it shouldn't, I just drew the line here to get initial version merged with support for most common cases.

chriadam commented on src/vk/vk-posts/vkpostsyncadaptor.cpp:
```diff
@@ -0,0 +1,472 @@
+                            ? QString::number(post.value(QStringLiteral("id")).toDouble())
+                            : QString::number(post.value(QStringLiteral("post_id")).toDouble()));
+
+    newPost.link = QStringLiteral("https://m.vk.com/wall") + identifier;
```
should that be https://m.vk.com/wall/ + identifier?  missing / separator?

antseppa replied:
no, the format really is wall123456789 without separator.

chriadam commented on src/vk/vk-posts/vkpostsyncadaptor.cpp:
```diff
@@ -0,0 +1,472 @@
+                            ? QString::number((int)post.value(QStringLiteral("id")).toDouble())
+                            : QString::number((int)post.value(QStringLiteral("post_id")).toDouble()));
+
+    newPost.link = QStringLiteral("https://m.vk.com/album")
```
again, do we need / separator?  or is this correct?

antseppa replied:
same as above

chriadam commented on src/vk/vk-posts/vkpostsyncadaptor.cpp:
```diff
@@ -0,0 +1,472 @@
+    // we want to use the largest of these dimensions as the "optimal"
+    int maxDimension = qMax(width, height);
+    if (maxDimension >= 2048) {
+        m_optimalImageSize = "photo_1280";
```
why 1280 if dimension is 2048?

antseppa replied:
That was the highgest the spec mentioned, but there actually is higher one in the server reply.

antseppa commented:
Updated:
- squashed to one commit
- adapted to use VKImageDatabase const ptr pattern
- fixed review comments

Still TODO:
- contacts sync adaptor needs some work:
     * filter out empty contacts
     * implement & test continuationUrl mechanism
     * make sure we won't exceed two requests-per-second limit with the above
- notifications sync adaptor purge

chriadam commented:
Not sure that squashing down to one commit is a good idea, but I guess it doesn't matter.

Commit:
Adapted to use ConstrPtr pattern with VKImageDatabase structs

antseppa commented:
Had the original structure still left, decided to unsquash after discussion. Otherwise same.

chriadam commented:
LGTM.  We can fix the issues with contact sync adaptor once we can test that one, I guess.

antseppa commented:
Updated: rebased

Commit:
Do not sync VK contacts without contact details

antseppa commented:
Updated: only sync those contacts which have contact details.

antseppa commented:
Updated: rebased

chriadam commented:
Still LGTM
